source/MaterialXGenGlsl/Nodes/HeightToNormalGlsl.cpp

```diff
@@ -5,237 +5,76 @@
-                                    context.removeOutputSuffix(upstreamOutput);
-
-                                    // Keep track of the output name with the suffix
-                                    sampleStrings.push_back(outputName + outputSuffix);
-                                }
-                            }
-                        }
-                        else
-                        {
-                            // Use the same input for all samples.
-                            // Note: This does not recomputed the output, but instead it reuses
-                            // the output variable.
-                            for (unsigned int i = 0; i < sampleCount; i++)
-                            {
-                                // On failure just call the unmodified function
-                                sampleStrings.push_back(outputName);
-                            }
-                        }
-                    }
-                }
-            }
-        }
-        else
-        {
-            if (!inInput->value)
-            {
-                throw ExceptionShaderGenError("No connection or value found on heighttonormal node '" + node.getName() + "'");
-            }
-        }
-
-        // Build a set of samples with constant values
-        if (sampleStrings.empty())
-        {
-            string inValueString = inInput->value->getValueString();
-            for (unsigned int i = 0; i < sampleCount; i++)
-            {
-                sampleStrings.push_back(inValueString);
-            }
-        }
-    }
+void HeightToNormalGlsl::emitFunctionCall(const SgNode& node, SgNodeContext& context, ShaderGenerator& shadergen, Shader& shader_)
+{
+    const SgInput* inInput = node.getInput("in");
+    const SgInput* scaleInput = node.getInput("scale");
 
-    void HeightToNormalGlsl::emitFunctionCall(const SgNode& node, SgNodeContext& context, ShaderGenerator& shadergen, Shader& shader_)
+    if (!inInput || !scaleInput)
     {
-        const SgInput* inInput = node.getInput("in");
-        const SgInput* scaleInput = node.getInput("scale");
-
-        if (!inInput || !scaleInput)
-        {
-            throw ExceptionShaderGenError("Node '" + node.getName() + "' is not a valid heighttonormal node");
-        }
+        throw ExceptionShaderGenError("Node '" + node.getName() + "' is not a valid heighttonormal node");
+    }
 
-        HwShader& shader = static_cast<HwShader&>(shader_);
+    HwShader& shader = static_cast<HwShader&>(shader_);
 
-        BEGIN_SHADER_STAGE(shader, HwShader::PIXEL_STAGE)
+    BEGIN_SHADER_STAGE(shader, HwShader::PIXEL_STAGE)
+    {
+        // Create the input "samples". This means to emit the calls to 
+        // compute the sames and return a set of strings containaing
+        // the variables to assign to the sample grid.
+        //  
+        _sampleStrings.clear();
+        emitInputSamplesUV(node, context, shadergen, shader);
```

Comment:
Convolutions will emit the input sample code first then use it later on.

Reply:
Does it really need to cache the sample results in a member (here the _sampleStrings)? Could the results instead be returned in an output vector? That is what emitBsdfNodes or emitEdfNodes does for example.
I'm just worried that if the code gen functions will change the state of the node implementation, it could bite us later in multi-threaded generations.

Reply:
I'll look at changing this. I originally had it as being passed back in a in/out arg.